exporter/metadataexporter/exporter.go

```diff
@@ -19,21 +34,385 @@ type tagValueCountFromDB struct {
+	tracesTagValueCountCtx       context.Context
+	tracesTagValueCountCtxCancel context.CancelFunc
+
+	alwaysIncludeTracesAttributes map[string]struct{}
+}
+
+func flattenJSONToStringMap(data map[string]any) map[string]string {
+	res := make(map[string]string, len(data))
+	for k, v := range data {
+		res[k] = fmt.Sprint(v)
+	}
+	return res
 }
 
 func newMetadataExporter(cfg Config, set exporter.Settings) (*metadataExporter, error) {
-	return &metadataExporter{cfg: cfg, set: set}, nil
+	opts, err := clickhouse.ParseDSN(cfg.DSN)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := clickhouse.Open(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	fingerprintCache := ttlcache.New[string, bool](
+		ttlcache.WithTTL[string, bool](300*time.Minute),
+		ttlcache.WithDisableTouchOnHit[string, bool](), // don't update the ttl when the item is accessed
+		ttlcache.WithCapacity[string, bool](3_000_000), // max 3M items in the cache
+	)
+	go fingerprintCache.Start()
+
+	tracesTracker := NewValueTracker(
+		int(cfg.MaxDistinctValues.Traces.MaxKeys),
+		int(cfg.MaxDistinctValues.Traces.MaxStringDistinctValues),
+		45*time.Minute, // if a key is not seen in 45 mins, it is removed from the tracker
+	)
+
+	tracesTagValueCountCtx, tracesTagValueCountCtxCancel := context.WithCancel(context.Background())
+
+	alwaysIncludeTraces := make(map[string]struct{}, len(cfg.AlwaysIncludeAttributes.Traces))
+	for _, attr := range cfg.AlwaysIncludeAttributes.Traces {
+		alwaysIncludeTraces[attr] = struct{}{}
+	}
+
+	// Initialize atomic pointers to empty maps.
+	initMap := func() *map[string]tagValueCountFromDB {
+		m := make(map[string]tagValueCountFromDB)
+		return &m
+	}
+
+	e := &metadataExporter{
+		cfg:                           cfg,
+		set:                           set,
+		conn:                          conn,
+		fingerprintCache:              fingerprintCache,
+		tracesTracker:                 tracesTracker,
+		tracesTagValueCountCtx:        tracesTagValueCountCtx,
+		tracesTagValueCountCtxCancel:  tracesTagValueCountCtxCancel,
+		alwaysIncludeTracesAttributes: alwaysIncludeTraces,
+	}
+
+	e.tracesTagValueCountFromDB.Store(initMap())
+
+	return e, nil
 }
 
 func (e *metadataExporter) Start(_ context.Context, host component.Host) error {
+	e.set.Logger.Info("starting metadata exporter")
+
+	go e.periodicallyUpdateTagValueCountFromDB(
+		e.tracesTagValueCountCtx,
+		&updateParams{
+			logger: e.set.Logger,
+			conn:   e.conn,
+			query: `SELECT tag_key, tag_data_type, countDistinct(string_value) as string_value_count, countDistinct(number_value) as number_value_count
+						 FROM signoz_traces.distributed_tag_attributes_v2
+						 GROUP BY tag_key, tag_data_type
+						 ORDER BY number_value_count DESC, string_value_count DESC, tag_key
+						 LIMIT 1 BY tag_key, tag_data_type`,
+			storeFunc:  e.storeTracesTagValues,
+			signalName: pipeline.SignalTraces.String(),
+			interval:   15 * time.Minute,
```

Comment:
this was to be 10 mins ?

Reply:
It has been 15 minutes in the test environment, we can probably revisit it later